Use custom roslyn/diagnostics endpoint if enabled

ryzngard opened the pull request:
The Razor side of https://github.com/dotnet/roslyn/pull/71365 

* [x] Still need manual testing. I hit some MEF things so I'm doing a purge of my cache etc, but putting up as I do so
* [x] Have an off switch, in case somehow we find edgecases that are ruining things for users. I assume this is better than an option?

davidwengier commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Endpoints/Diagnostics.cs:
```diff
@@ -74,19 +76,41 @@ internal partial class RazorCustomMessageTarget
+            var diagnosticParams = new RazorDiagnosticsParams
+            {
+                RazorTextDocument = identifierFromOriginalRequest,
+                TextDocument = identifierFromOriginalRequest.WithUri(virtualDocument.Uri),
```
This won't work, I believe. `WithUri` modifies the original TextDocumentIdentifier, so both of these properties would end up pointing to the same Uri. 

Having said that, a `new TextDocumentIdentifier { Uri = virtualDocument.Uri }` should be sufficient any way, as the project context info that might be lost is only relevant to us, not Roslyn, and the virtual document Uris are unique per project anyway.

ryzngard replied:
I believe this to be fixed

davidwengier commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Endpoints/Diagnostics.cs:
```diff
@@ -74,19 +76,41 @@ internal partial class RazorCustomMessageTarget
-        };
-
         var lspMethodName = VSInternalMethods.DocumentPullDiagnosticName;
         using var _ = _telemetryReporter.TrackLspRequest(lspMethodName, delegatedLanguageServerName, correlationId);
```
This probably needs to move, or we'll be mixing up telemetry.

ryzngard commented:
/azp run

ryzngard commented on SpellingExclusions.dic:
```diff
@@ -3,3 +3,4 @@ Blazor
 csproj
 cshtml
 microsoft
+csharp
```
I'm not sure... it might be? I'll revert. I just did the automatic fix on the file I was editing

ryzngard commented:
@davidwengier would you mind doing another pass? I added some more logic for the mapping portion that was missing and updated some tests. I think it warrants another look

maryamariyan commented on src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/VisualStudioWindowsLanguageServerFeatureOptions.cs:
```diff
@@ -51,6 +53,13 @@ public VisualStudioWindowsLanguageServerFeatureOptions(LSPEditorFeatureDetector
+        _useRoslynAnalyzerDiagnostics = new Lazy<bool>(() =>
+        {
+            var featureFlags = (IVsFeatureFlags)AsyncPackage.GetGlobalService(typeof(SVsFeatureFlags));
+            var usePreciseSemanticTokenRanges = featureFlags.IsFeatureEnabled(UseRoslynAnalyzerDiagnosticsFeatureFlag, defaultValue: true);
```
nit: seem to be using outdated variable name `usePreciseSemanticTokenRanges` from copy paste

davidwengier commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/RazorTranslateDiagnosticsService.cs:
```diff
@@ -479,6 +480,18 @@ private bool TryGetOriginalDiagnosticRange(Diagnostic diagnostic, RazorCodeDocum
+                if (diagnostic.Range.Start.TryGetAbsoluteIndex(sourceText, _logger, out var _) &&
+                    diagnostic.Range.End.TryGetAbsoluteIndex(sourceText, _logger, out var _))
+                {
+                    // We can't remap the range correctly, but it's probably an analyzer diagnostic so we still want to show it to the user
+                    // if it's still in the razor document range
```
This is a very scary assumption. In a non-trivial razor file I would expect a lot of false positives here from errors in generated code that happens to be within the length of a razor document. Given we know of exactly one scenario that exists where there is actually a razor analyzer, and its not really a supported scenario, I would guess that most of the time this will be hit will actually be a false positive.

Can we tag the diagnostics in some way, with a custom tag that indicates if its from the .razor file and not the .razor.g.cs file? eg, based on the `identifier` parameter that is provided here: https://github.com/dotnet/roslyn/pull/71365/files#diff-7eaffa27e852935ff999303c53260d548c4a45c4f6e9901740df320dfcd229dfR33

_HOWEVER_ the reason I say that is because I don't know what a "Compiler" DiagnosticType implies. Does it, in fact, act as a way of tagging which diagnostics come from the Razor analyzer? What do IDE and CA diagnostics in the genereated file come back as?

ryzngard replied:
I need to check those, good call out. Will get to it in the pst am

The type is vs specific and compiler is hard code for cs###

DustinCampbell commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/RazorTranslateDiagnosticsService.cs:
```diff
@@ -495,6 +508,19 @@ private bool TryGetOriginalDiagnosticRange(Diagnostic diagnostic, RazorCodeDocum
         return true;
     }
 
+    private bool IsCompilerDiagnostic(Diagnostic diagnostic)
```
nit: looks like this helper can be static.

ryzngard commented:
This is superseded by the Roslyn compiler fix https://github.com/dotnet/roslyn/pull/71454
